File: solution/java/week06/exercise2/src/test/java/report/ReportGeneratorTests.java
package report;

import org.approvaltests.Approvals;
import org.junit.jupiter.api.AfterEach;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;

import java.io.ByteArrayOutputStream;
import java.io.PrintStream;
import java.util.List;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertTrue;

public class ReportGeneratorTests {

    private final ByteArrayOutputStream outContent = new ByteArrayOutputStream();
    private final PrintStream originalOut = System.out;

    @BeforeEach
    public void setUpStreams() {
        System.setOut(new PrintStream(outContent));
    }

    @AfterEach
    public void restoreStreams() {
        System.setOut(originalOut);
    }

    @Test
    public void generateReportInCSV() {
        List<ReportData> data = List.of(
                new ReportData(1, 100.0, "Sample Data 1"),
                new ReportData(2, 200.0, "Sample Data 2")
        );

        var generator = new ReportGenerator();
        var report = generator.generateReport(ReportType.CSV, data);
        report.onSuccess(System.out::println)
                .onFailure(System.err::println);

        Approvals.verify(outContent.toString());
    }

    @Test
    public void generateReportInPDF() {
        List<ReportData> data = List.of(
                new ReportData(1, 100.0, "Sample Data 1"),
                new ReportData(2, 200.0, "Sample Data 2")
        );

        var generator = new ReportGenerator();
        var report = generator.generateReport(ReportType.PDF, data);
        report.onSuccess(System.out::println)
                .onFailure(System.err::println);

        Approvals.verify(outContent.toString());
    }

    @Test
    public void generateReportInPDFShouldFailIfReportDataIsEmpty() {
        var generator = new ReportGenerator();
        var report = generator.generateReport(ReportType.PDF, null);
        report.onSuccess(System.out::println)
                .onFailure(System.err::println);

        assert report.isFailure();
        report.onFailure(ex -> {
            assertTrue(ex instanceof NoDataToReportException, "Expected a NoDataToReportException, but got " + ex.getClass().getSimpleName());
            assertEquals("No data provided for report generation.", ex.getMessage());
        });
    }

    @Test
    public void generateReportInPDFShouldFailIfReportTypeNotSupported() {
        List<ReportData> data = List.of(
                new ReportData(1, 100.0, "Sample Data 1"),
                new ReportData(2, 200.0, "Sample Data 2")
        );

        var generator = new ReportGenerator();
        var report = generator.generateReport(ReportType.Unsupported, data);
        report.onSuccess(System.out::println)
                .onFailure(System.err::println);

        assert report.isFailure();
        report.onFailure(ex -> {
            assertTrue(ex instanceof UnsupportedReportTypeException, "Expected a NoDataToReportException, but got " + ex.getClass().getSimpleName());
            assertEquals("Report type Unsupported not supported.", ex.getMessage());
        });
    }
}

File: solution/c#/week06-legacy-code/ReportGenerator.Tests/ReportGeneratorTests.cs
using FluentAssertions;
using FluentAssertions.LanguageExt;
using LanguageExt;
using LanguageExt.UnsafeValueAccess;

namespace ReportGenerator.Tests
{
    public class ReportGeneratorTests
    {
        [Fact]
        public Task Generate_Report_In_Csv() =>
            Verify(
                CreateReportGenerator()
                    .GenerateReport(ReportType.Csv, CreateReportData())
                    .ValueUnsafe()
            );

        [Fact]
        public Task Generate_Report_In_Pdf() =>
            Verify(
                CreateReportGenerator()
                    .GenerateReport(ReportType.Pdf, CreateReportData())
                    .ValueUnsafe()
            );

        [Fact]
        public void Generate_Report_Should_Fail_If_Report_Data_Is_Empty() =>
            CreateReportGenerator()
                .GenerateReport(ReportType.Pdf, [])
                .Should()
                .BeLeft(error => error
                    .Should()
                    .Be(new NoDataToReport("No data provided for report generation.")
                    ));

        [Fact]
        public void Generate_Report_Should_Fail_If_Report_Type_Not_Supported() =>
            CreateReportGenerator()
                .GenerateReport(ReportType.Unsupported, CreateReportData())
                .Should()
                .BeLeft(error => error
                    .Should()
                    .Be(new UnsupportedReportType("Report type Unsupported not supported.")
                    ));

        private static ReportGenerator CreateReportGenerator() =>
            new(new Dictionary<ReportType, ICanGenerateReport>
            {
                {ReportType.Csv, new CsvReportGenerator()},
                {ReportType.Pdf, new PdfReportGenerator()}
            });

        private static Seq<ReportData> CreateReportData() =>
            new List<ReportData>
            {
                new(1, 100, "Sample Data 1"),
                new(2, 200, "Sample Data 2")
            }.ToSeq();
    }
}